### PR TITLE
[ty] Use TDD-based narrowing constraints and support NoReturn narrowing

### PR DESCRIPTION
**Pull request by alex**
Replace sorted-list-based narrowing constraints (AND-only) with TDD (Ternary Decision Diagram) nodes that support AND, OR, and NOT. This changes the merge operation from intersection to OR, so narrowing from non-terminal branches is properly preserved after if/elif/else.

Additionally, when a branch contains a NoReturn call (e.g. sys.exit()), gate the narrowing constraints by the ReturnsNever predicate. During the narrowing TDD walk, non-narrowing predicates are evaluated to determine reachability, so NoReturn branches contribute Never (absorbed by union) rather than polluting the merged type.

Fixes https://github.com/astral-sh/ty/issues/690
Fixes https://github.com/astral-sh/ty/issues/685

### CODE REVIEW

**Comment by alex** [comment]
(Probably not ready for review yet, just wanted to see the ecosystem results.)

**Comment by alex** [comment]
Hard to know exactly how to verify the mypy_primer results, but they appear to be overwhelmingly getting rid of "no attribute on X | None" which is basically what you'd expect form better narrowing propagation.

**Comment by carljm** [comment]
> Hard to know exactly how to verify the mypy_primer results, but they appear to be overwhelmingly getting rid of "no attribute on X | None" which is basically what you'd expect form better narrowing propagation.

We've currently got some non-determinism we're working to fix, so the mypy-primer results are particularly difficult to parse right now -- some projects (sympy, prefect) will have a lot of totally unrelated changes in the diffs. The results look very good to me overall.

**Comment by carljm** [comment]
> (Probably not ready for review yet, just wanted to see the ecosystem results.)

You can also open a draft PR and you'll still get the CI results. Lmk if you want an early review on this now, if not maybe mark it draft until you do.

Super exciting, thank you for taking this on!

**Comment by alex** [comment]
An early review would be appreciated, thanks! Particularly if there's test cases you think it's missing.

**Comment by sharkdp** [comment]
This looks really great, thank you!

I only did a coarse review as I understand that this is not quite finished.

It's fantastic to see that we can make use of the existing TDD for narrowing constraints as well. I think we should eventually rename some types/functions (reachability constraints => some new term), if we go through with this, but it's probably better to do this in a follow-up to keep the diff small.

> Particularly if there's test cases you think it's missing.

One test case that I found that is still not handled with this PR is the following:
```py
def _(x: int | None):
    if 1 + 1 == 2:
        if x is None:
            return
        reveal_type(x)  # revealed: int

    reveal_type(x)  # revealed: `int | None`, should ideally be `int`
```
Here, `1 + 1 == 2` is a placeholder for any nontrivial condition that evaluates to `Literal[True]`, like `sys.version_info >= (3, 9)` on a recent Python version. If you replace it with `True`, it works fine. Handling this would probably involve adding new nodes to the narrowing constraints TDD that would later evaluate the truthiness of the test condition? This seems like a rather obscure edge case, so I'm not sure if it's worth adding a lot of complexity.

**Comment by alex** [comment]
Interesting, in general those sorts of compile time values feel like they though to be basically the same as `Literal[True]` for all purposes. I'll see what it'd take to implement (depending on what it looks like, I might conclude one of: a) it proves that some of the abstractions in this PR are wrong, b) it's easy, just include, c) it's still annoying, defer to follow up).

**Comment by carljm** [comment]
> Interesting, in general those sorts of compile time values feel like they though to be basically the same as `Literal[True]` for all purposes.

That did used to be true in ty (they were all handled via the reachability-constraints mechanism), but then we added special-cased earlier handling of "simple" statically-known conditions that don't require type inference to resolve. We did that because it was a significant performance boost -- but it does mean that it's now easy to accidentally introduce different behavior for the two cases.

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md [hunk 1]
```diff
@@ -49,3 +49,154 @@ def _(x: int | None):
...
+    if isinstance(x, A):
+        pass
+    elif isinstance(x, B):
+        pass
+    else:
+        return
+
+    # Only the if-branch (A) and elif-branch (B) flow through.
+    # The else-branch returned, so its narrowing doesn't participate.
+    reveal_type(x)  # revealed: B | (A & ~B)
```
Comment (carljm) [inline crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md:73]: I'm a bit confused about why this isn't `A | (B & ~A)`. If we flow through the `if` we should just get `A`, if we flow through the `elif` we have `B & ~A` -- adding `reveal_type(x)` inside the branches above confirms. So it's not clear to me why we don't end up with the union of those here.

I guess those are equivalent types (and also equivalent to just `A | B`?) -- so maybe this is an acceptable (if slightly surprising) transformation that happens due to the TDD implementation?

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md [hunk 2]
```diff
@@ -49,3 +49,154 @@ def _(x: int | None):
...
+        return
+    elif isinstance(x, B):
+        pass
+    elif isinstance(x, C):
+        pass
+    else:
+        return
+
+    # Only the elif-B and elif-C branches flow through.
+    reveal_type(x)  # revealed: (C & ~A) | (B & ~A & ~C)
```
Comment (carljm) [inline crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md:95]: Similarly here I'd naively expect `(B & ~A) | (C & ~A & ~B)` -- but they are equivalent. I'm just curious why it happens.

**Review by carljm — COMMENTED** [review]
Similar to  @sharkdp 's suggested test, but slightly different (though I suspect with the same fix) is the OP example in https://github.com/astral-sh/ty/issues/685 but modified to use a condition that (currently) requires type inference:

```py
def _(val: int | None):
  if val is None:
    if 1+1+2:
      return

  reveal_type(val) # revealed: int | None, should be just `int`
```

(Of course "requires type inference" is slippery here -- we could certainly fix this by implementing direct support for more constexprs like this in semantic indexing. The real example here would be something like importing a conditionally-defined `Final` constant from another module; we can't resolve that in semantic indexing, we need a reachability constraint to be resolved later.)

**Comment by carljm** [comment]
We could also add some tests using `match` statements; one example (could be simplified) in https://github.com/astral-sh/ty/issues/1384

**Comment by carljm** [comment]
Took a pass on the tests -- going to have @sharkdp take the lead on review here.

**Comment by AlexWaygood** [comment]
Not sure why the ecosystem-analyzer comment hasn't updated, but the latest run shows we're now timing out on 4 ecosystem projects unfortunately (sympy, setuptools, jax and black): https://github.com/astral-sh/ruff/actions/runs/21756978896?pr=23109

**Comment by alex** [comment]
Yup -- that wasn't a problem before adding https://github.com/astral-sh/ruff/pull/23109/changes/ada1084d84612f15773701c83f1ffb8f0733519b -- looking to see if there's some expontential blow up, an infinite loop, or something else.

**Comment by woodruffw** [comment]
I poked at this, and I think it was a flake on CloudFlare's end. I've kicked the workflow over and this time it should update correctly, but if not I'll triage further.

**Comment by alex** [comment]
There's an exponential blow up issue (16 nested ifs in black). I don't see an obvious way to address directly, so I'm going to cap the nesting for this analysis. Which is a bit sad from a generality perspective 🙃

**Comment by alex** [comment]
Ok, ended up rolling back the fix for `if 1+1 == 2` propagation due to exponential blow up.

**Comment by alex** [comment]
I think this is ready for review:

1. The implementation I had for solving the `if 1 + 1 == 2` branching issue had serious performance problems. Conceptually I feel like this should be possible (the branch is statically known, that's the point!), but given the obvious impl didn't work, I think it makes more sense to defer.
2. There's still some smaller performance regressions on 3 of the benchmarks. I don't feel well calibrated for to what extent that's expected and/or acceptabled.
3. My review of the mypy_primer results looks ok. I can't say that I scruitinized every single one, but I skimmed the list and looked at any that seemed interesting/questionable and they all looked either correct, or like an orthogonal issue.
4. It does seem like some sort of normalization would be valuable (e.g., `(A & ~B) | (B & ~A)` really should just be `A | B`), but I think that's basically a consequence of the existing TDD impl/intersection builder, and thus shouldn't be part of this PR. Let me know if you disagree.

**Comment by AlexWaygood** [comment]
> 2\. There's still some smaller performance regressions on 3 of the benchmarks. I don't feel well calibrated for to what extent that's expected and/or acceptabled.

I think that's within the acceptable range personally, but I'm a bit surprised, because the original version of this PR showed a much smaller regression on codspeed IIRC, and I thought the idea of https://github.com/astral-sh/ruff/pull/23109/commits/5fd9a9c27056d55a62a138c549699e0b338d7bb7 was to go back to the original version of this PR

**Comment by AlexWaygood** [comment]
> `(A & ~B) | (B & ~A)` really should just be `A | B`

I'm not sure that's actually true — the first type excludes any runtime values that are instances of both `A` and `B` (which can happen due to multiple inheritance), whereas the second type does not (it's a larger set).

**Comment by alex** [comment]
Ugh, multiple inheritance.

`git diff 2f7e96e59b02b9e1f4b350b27b8e5a47ddb5f53f` is test-only at this point, so if there's a performance difference against that commit it may be variability (ugh).

**Comment by carljm** [comment]
It is true that `A | (B & ~A)` is equivalent to `A | B` though.

Re perf, it may be that the current non determinism in our results on some projects spills over into variability in the benchmarks, since eg emitting diagnostics can be expensive.

**Comment by alex** [comment]
Hmm, I think that's not strictly true?

```py
class A: ...
class B: ...
class C(A, B): ...

x: A | (B & ~A)
if isintance(x, B):
    # x can't be C here, because then it'd be A, which we know it's not.
    # If we simplify x to A | B, then we lose that constraint
```

**Comment by carljm** [comment]
But that constraint is not correct, because nothing prevents a C from inhabiting the `A` side of the union and passing `isinstance(x, B)`. The type expressed by the `B` side of the union does not have an implicit monopoly on instances of `B` in the overall union. 

If you [try that example in ty](https://play.ty.dev/49772ebb-f2ca-4feb-83df-9151911261f2), it will tell you that the type of `x` inside the `if` block is `(A & B) | (B & ~A)`, expressing that possibility that a `C` could inhabit the `A` side of the union. And this type is equivalent to just `B`, which is what we'd get if we simplified the original union to `A | B`. 

This is why `A | (B & ~A)` simplifies but `(A & ~B) | (B & ~A)` does not; the latter _does_ exclude `C` from the `A` side of the union.

**Comment by alex** [comment]
Gah, too early on a Saturday for multiple inheritance apparently :-( Thanks.

**Comment by TomerBin** [comment]
Amazing 😍 
Can't wait to keep working on https://github.com/astral-sh/ty/issues/626 once this is landed

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md (hunk 1 above)
Comment (sharkdp) [reply crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md:73]: > I guess those are equivalent types (and also equivalent to just `A | B`?) -- so maybe this is an acceptable (if slightly surprising) transformation that happens due to the TDD implementation?

Yes, the root cause of this is the reversed ordering of TDD nodes that we introduced in https://github.com/astral-sh/ruff/pull/20098.

This didn't have any observable side effects for reachability constraints (which evaluate to a ternary answer), but it does play a role for narrowing constraints.

Removing that `.reverse()` changes the result here to `A | (B & ~A)`. But it also breaks ~20 other tests. Some of them are just ordering related, but in some other cases, some union-builder simplifications do not seem to trigger which is a bit concerning. I have not investigated more closely, as we probably still need that optimization.

File: crates/ty_python_semantic/src/semantic_index/use_def.rs [hunk 3]
```diff
@@ -1017,22 +984,53 @@ impl<'db> UseDefMapBuilder<'db> {
...
+        let atom = self.reachability_constraints.add_atom(predicate);
+        self.record_narrowing_constraint_node_for_places(atom, places);
+    }
+
+    /// Records a negated narrowing constraint for only the specified places.
+    ///
+    /// Uses TDD-level negation (`add_not_constraint`) rather than creating a new predicate atom
+    /// for the negated predicate. This ensures that `atom(P) OR NOT(atom(P))` simplifies to
+    /// `ALWAYS_TRUE` in the TDD, so narrowing is correctly cancelled out after complete
+    /// if/else blocks.
```
Comment (sharkdp) [inline crates/ty_python_semantic/src/semantic_index/use_def.rs:996]: So this makes me wonder if we can completely replace the `is_positive` flag in predicates and replace it with `NOT(…)` from the TDD? With the change here, we now use `P` with `is_positive: false` for some situations and `NOT(P)` for other situations.

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md [hunk 4]
```diff
@@ -49,3 +49,192 @@ def _(x: int | None):
...
+```
+
+## Multiple sequential if-statements don't leak narrowing
+
+After a complete if/else where both branches flow through (no terminal), narrowing should be
+cancelled out at the merge point.
+
+```py
+def _(x: int | str | None):
```
Comment (sharkdp) [inline crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md]: Minor: Can we use `A, B, C` for this example as well? Just like in the tests above.

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md [hunk 5]
```diff
@@ -49,3 +49,192 @@ def _(x: int | None):
...
+    else:
+        pass
+
+    # Narrowing cancels out: both paths flow, so type is unchanged.
+    reveal_type(x)  # revealed: int | str | None
+
+    if isinstance(x, str):
+        y = 1
+    else:
+        y = 2
```
Comment (sharkdp) [inline crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md]: I found `y` distracting/confusing here

```suggestion
    if isinstance(x, str):
        pass
    else:
        pass
```

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md [hunk 6]
```diff
@@ -49,3 +49,192 @@ def _(x: int | None):
...
+        reveal_type(x)  # revealed: int
+
+    # TODO: should be `int` (the else-branch of `1 + 1 == 2` is unreachable)
+    reveal_type(x)  # revealed: int | None
+```
+
+This also works when the always-true condition is nested inside a narrowing branch:
+
+```py
+def _(val: int | None):
```
Comment (sharkdp) [inline crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md]: For symmetry/readability, also use `x` here? Makes it easier to spot the difference between this test and the one above.

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md [hunk 7]
```diff
@@ -49,3 +49,192 @@ def _(x: int | None):
...
+    return None
+
+def bar() -> None:
+    v = foo(1)
+    assert v is None
+
+    v = foo(2)
+    # v was reassigned, so the assert narrowing shouldn't apply
+    reveal_type(v)  # revealed: int | None
+```
```
Comment (sharkdp) [inline crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md:222]: This is not a post-if-statement test. I would think that we have similar tests elsewhere, or is this a regression test for something that happened during development?

File: crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs [hunk 8]
```diff
@@ -734,7 +735,178 @@ impl ReachabilityConstraintsBuilder {
...
+    ///
+    /// At each interior node, we branch based on whether the predicate is true or false:
+    /// - True branch: apply positive narrowing from the predicate
+    /// - False branch: apply negative narrowing from the predicate
+    ///
+    /// The "ambiguous" branch in the TDD is not followed for narrowing purposes, because
+    /// narrowing constraints record which predicates hold along the control flow path.
+    /// The predicates may be statically ambiguous (we can't determine their truthiness
+    /// at analysis time), but they still hold dynamically at runtime and should be used
+    /// for narrowing.
```
Comment (sharkdp) [inline crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs:779]: I think this means that a BDD would in principle be sufficient for the narrowing constraints?

File: crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs [hunk 9]
```diff
@@ -734,7 +735,178 @@ impl ReachabilityConstraintsBuilder {
...
+                if pos_constraint.is_none() {
+                    // This predicate doesn't narrow our variable (e.g., it's a
+                    // `ReturnsNever` constraint, or a narrowing constraint for a
+                    // different variable). Evaluate it to determine which path(s)
+                    // are reachable, rather than walking both branches.
+                    let neg_predicate = Predicate {
+                        node: predicate.node,
+                        is_positive: !predicate.is_positive,
+                    };
+                    let neg_constraint = infer_narrowing_constraint(db, neg_predicate, place);
```
Comment (sharkdp) [inline crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs:860]: Why do we have to check the negated version of the predicate as well? In which case does a constraint return `None` from `infer_narrowing_constraint`, but its negated version returns `Some(…)`? And why do we need to guard against that here?

Comment (sharkdp) [reply crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs:860]: Oh, I guess this is relevant if `pos_predicate` has `is_positive: false`? In that case, can we maybe just set `is_positive: true` and do just one check?

In any case, I would appreciate an explanation here why we need to consider both variants.

Comment (sharkdp) [reply crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs:860]: Thinking more about this… this seems like a very wasteful way of detecting `NoReturn` predicates. Can't we simply pattern match on `predicate.node` to get the same information?

Doing this check instead of the two `{pos,neg}_constraint.is_none()` checks also seems to work:
```rs
if matches!(predicate.node, PredicateNode::ReturnsNever(_)) {
```

File: crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs [hunk 10]
```diff
@@ -734,7 +735,178 @@ impl ReachabilityConstraintsBuilder {
...
+                                );
+                                let false_ty = self.narrow_by_constraint_inner(
+                                    db,
+                                    predicates,
+                                    node.if_false,
+                                    base_ty,
+                                    place,
+                                    accumulated,
+                                );
+                                UnionType::from_elements(db, [true_ty, false_ty])
```
Comment (sharkdp) [inline crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs]: I don't understand this. What does it mean to be in this branch, and why do we want to return the union of true_ty and false_ty?

Can this branch even be reached? `analyze_single` for a `ReturnsNever` branch can only ever return `AlwaysTrue` or `AlwaysFalse`.

File: crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs [hunk 11]
```diff
@@ -734,7 +735,178 @@ impl ReachabilityConstraintsBuilder {
...
+                let false_ty = self.narrow_by_constraint_inner(
+                    db,
+                    predicates,
+                    node.if_false,
+                    base_ty,
+                    place,
+                    false_accumulated,
+                );
+
+                UnionType::from_elements(db, [true_ty, false_ty])
```
Comment (sharkdp) [inline crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs:913]: It feels like this would benefit from some fast-paths for `AlwaysFalse` nodes? Otherwise, we'll build a lot of unions with `Never`.

**Review by sharkdp — COMMENTED** [review]
Thank you very much for working on this.

I feel like I now have an understanding of how this approach works, so I'm submitting a first review with some higher-level questions, before doing another final pass.

File: crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md (hunk 7 above)
Comment (alex) [reply crates/ty_python_semantic/resources/mdtest/narrow/post_if_statement.md:222]: Yeah, this was a regression I hit in the _first_ version of this I worked on a month or so ago.

File: crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs (hunk 8 above)
Comment (alex) [reply crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs:779]: I think so, we're just piggy backing on the TDD from the constraints.

File: crates/ty_python_semantic/src/semantic_index/use_def.rs (hunk 3 above)
Comment (alex) [reply crates/ty_python_semantic/src/semantic_index/use_def.rs:996]: I think yes, it is. But I don't want to make this PR too large, it's already annoyingly large :-( And I don't think the inverse is true, I don't think we could use is_positive for this.

**Comment by alex** [comment]
Is the memory usage report new, or did it have no significant results before?

**Comment by ibraheemdev** [comment]
That is new, but the numbers look fine, I wouldn't worry about it.

File: crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs (hunk 11 above)
Comment (sharkdp) [reply crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs:913]: Looks like this helped a lot with performance — all regressions are basically gone now.

File: crates/ty_python_semantic/src/semantic_index/use_def.rs (hunk 3 above)
Comment (sharkdp) [reply crates/ty_python_semantic/src/semantic_index/use_def.rs:996]: Ok, I'll look into this in a follow-up.

**Comment by alex** [comment]
Whoo! Thanks so much!
